barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -622,3 +622,72 @@ namespace avm_alu(256);
     #[SHL_OUTPUT]
     op_shl * (ic - (b_lo * two_pow_s * shift_lt_bit_len)) = 0;
    
+    // ========= INTEGER DIVISION ===============================
+    // Operands: ia contains the dividend, ib contains the divisor, and ic contains the quotient (i.e. the result).
+    // All operands are restricted to be U128.
+    // The logic for integer division is to assert the correctness of this relationship:
+    // dividend - remainder = divisor * quotient ==> ia - remainder = ib * ic; where remainder < ib
+    // We do this using the following steps
+    // (1) The only non-trivial division is the situation where ia > ib && ib > 0
+    //      (a) if ia == ib => ic = 1 and remainder = 0 --> we can handle this as part of the standard division
+    //      (b) if ia < ib  => ic = 0 and remainder = ia --> isolating this case eliminates the risk of ia - remainder underflowing as remainder < ib < ia
+    //      (c) if ib == 0  => error_tag = 1 --> Handled in main trace
+    // (2) Given ib and ic are restricted to U128, at most ib * ic will produce a 256-bit number.
+    // (3) We use the primality check from cmp to check that this product has not overflowed the field
+    //     The Primality check takes a field element as input and ouputs two 128-bit limbs.
+    //     i.e. it checks that the field element, represented over 256-bits, is between [0, p).
+    //      (a) Given x, PC(x) -> [x_lo, x_hi], where x_lo < 2**128 && x_hi < 2**128 && x == x_lo + x_hi * 2**128
+    //      (b) Additionally produces a witness that the x < p
+    //          p_sub_x_lo = p_lo - x_lo + borrow * 2**128  < 2**128
```

Comment:
Here, p_lo is meant to be the lowest 128-bit limb of p-1, right? (and not p) If this is the case, it is worth to mention it.